contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,259 @@
+
+When working on a `feature`/`fix` branch, you may
+- [commit non-functioning code](#always-commit-functional-code), or
+- may need [multiple commits to implement one logical change in code](#one-logical-change-to-one-commit), or
+- cherry-pick a lot of commits from other branches
+
+This may clutter your history with low-level details or make it convoluted to follow for a maintainer or reviewer. Like sausage, you may enjoy eating it but not the process of making it.
+
+### How to Hide the Sausage
+Hiding the sausage is typically achieved by either/both:
+- performing an interactive rebase, i.e. `git rebase -i`
+- patch-wise reset and stage, i.e. `git reset -p` and `git add -p`
+
+Ensure that you do this _before_
+- pushing to a remote repository [to respect the published history](#respect-published-history) and
+- performing any merges from other branches.
+
+Read more:
+- [On Sausage Making](https://sethrobertson.github.io/GitBestPractices/#sausage)
+- [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) (great reference for how to actually do this, under _Changing Multiple Commit Messages_)
+
+---
+
+## Respect Published History
+Always avoid rewriting the public, published history unless you have a very good reason, you are very sure of what you are doing, or you will not affect anyone else.
+
+### Exceptions
+You are working on your own branch that _no one else is using_, and
+- you want to revert a commit without introducing another commit
+- you are rebasing the branch
+- you are [cleaning up the history of your branch](#hide-the-sausage-making)
+
+Read more:
+- [Rewriting History - Atlassian](https://www.atlassian.com/git/tutorials/rewriting-history)
+- [Don't change public history.](https://sethrobertson.github.io/GitBestPractices/#pubonce)
+
+---
+
+## Keep Up To Date
+### Use Remotes Effectively
+_Remotes_ refer to versions of the project you are working on that are hosted elsewhere, usually on the Internet. Remotes are very handy for managing collaboration, e.g. if you have to keep your code in sync with the `upstream` branch of the project, or if you need to pull code from someone else which may not be merged yet.
+
+You can have as many remotes as you want, each possibly being read-only or with read/write privileges.
+
+Remotes are managed using the `git remote` command.
+```bash
+# view all remotes
+git remote -v
+
+# add a remote called "upstream" that points to https://github.com/TEAMMATES/teammates
+git remote add upstream https://github.com/TEAMMATES/teammates
+
+# branch off from the master branch of the upstream repository
+git fetch upstream # get data from upstream repo
+git checkout -b your-fancy-branch upstream/master # makes a new branch off the head of upstream/master
+```
+
+Read more:
+- [Git Basics - Working with Remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes)
+
+### Rebase versus Merging
+| You should... | when ... |
```

Comment:
To standardize the headers, `when ...` should have the `w` capitalized and the space removed between `when` and `...` like so:
**W**hen**...**